magefiles/magefile.go

```diff
@@ -0,0 +1,389 @@
+			}
+
+			// Generate
+			if err = sh.RunV("protoc", "--twirp_out", ".", "--twirp_opt", "paths=source_relative",
+				"--go_out", ".", "--go_opt", "paths=source_relative", file); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	// It is called on the host
+	if err := sh.RunV("bash", "-c", "docker build -t trivy-protoc - < Dockerfile.protoc"); err != nil {
+		return err
+	}
+	return sh.Run("docker", "run", "--rm", "-it", "-v", "${PWD}:/app", "-w", "/app", "trivy-protoc", "mage", "protoc")
+}
+
+// Yacc generates parser
+func Yacc() error {
+	mg.Deps(Tool{}.Goyacc)
+	return sh.Run("go", "generate", "./pkg/licensing/expression/...")
+}
+
+// Easyjson generates JSON marshaler/unmarshaler for TinyGo/WebAssembly as TinyGo doesn't support encoding/json.
+func Easyjson() error {
+	mg.Deps(Tool{}.EasyJSON)
+	return sh.Run("easyjson", "./pkg/module/serialize/types.go")
+}
+
+type Test mg.Namespace
+
+// FixtureContainerImages downloads and extracts required images
+func (Test) FixtureContainerImages() error {
+	mg.Deps(Tool{}.Crane)
+	if err := os.MkdirAll(filepath.Join("integration", "testdata", "fixtures", "images"), 0750); err != nil {
+		return err
+	}
+
+	downloadScript := filepath.Join("integration", "scripts", "download-images.sh")
+	return sh.Run(downloadScript)
+}
+
+// FixtureVMImages downloads and extracts required VM images
+func (Test) FixtureVMImages() error {
+	mg.Deps(Tool{}.Crane)
+	downloadScript := filepath.Join("integration", "scripts", "download-vm-images.sh")
+	return sh.Run(downloadScript)
+}
+
+// GenerateModules compiles WASM modules for unit tests
+func (Test) GenerateModules() error {
+	pattern := filepath.Join("pkg", "module", "testdata", "*", "*.go")
+	if err := compileWasmModules(pattern); err != nil {
+		return err
+	}
+	return nil
+}
+
+// GenerateExampleModules compiles example Wasm modules for integration tests
+func (Test) GenerateExampleModules() error {
+	pattern := filepath.Join("examples", "module", "*", "*.go")
+	if err := compileWasmModules(pattern); err != nil {
+		return err
+	}
+	return nil
+}
+
+func compileWasmModules(pattern string) error {
+	goFiles, err := filepath.Glob(pattern)
+	if err != nil {
+		return err
+	}
+
+	for _, src := range goFiles {
+		// e.g. examples/module/spring4shell/spring4shell.go
+		//   => examples/module/spring4shell/spring4shell.wasm
+		dst := strings.TrimSuffix(src, ".go") + ".wasm"
+		if updated, err := target.Path(dst, src); err != nil {
+			return err
+		} else if !updated {
+			continue
+		}
+		// Check if TinyGo is installed
+		if !installed("tinygo") {
+			return errors.New("need to install TinyGo, follow https://tinygo.org/getting-started/install/")
+		}
+		if err = sh.Run("go", "generate", src); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Unit runs unit tests
+func (t Test) Unit() error {
+	mg.Deps(t.GenerateModules)
+	return sh.RunWith(ENV, "go", "test", "-v", "-short", "-coverprofile=coverage.txt", "-covermode=atomic", "./...")
+}
```

Comment:
can we add logs, like for make file?
```
➜  mage test:unit
Error: running "go test -v -short -coverprofile=coverage.txt -covermode=atomic ./..." failed with exit code 1

```

Reply:
Same for other tests

Reply:
`mage -v` shows stdout, but I forced it for tests.
https://github.com/aquasecurity/trivy/pull/3932/commits/f325cc61c234172cd96036c1989d383a2ddd2a1c